### DISCUSSION TITLE
Best way to parse the user_agent property?

### DISCUSSION THREAD

**Discussion by SuperSajuuk** [pallets/flask#4228] [Q&A] [answered]
Hi all, I'm looking to do some work with the user_agent property on a request object [I don't subclass Request, but use the default implementation]. I noticed in the documentation that this only contains a string property that must be parsed and online, I found a simple solution to parse that with a subclassed UserAgent. However, the example also suggested I needed to subclass Request to use that subclass.

Does anyone know how I would use the default user_agent property _without_ requiring the Request to be subclassed in order to use a custom parser for the user agent? Thank you! ❤

**Comment by tachyondecay** [comment, marked as answer]
You don't need to subclass `Request`, but you do need to subclass `werkzeug.user_agent.UserAgent`. Then you can set `app.request_class.user_agent_class` to be your new user agent class, like so:

```py
from flask import Flask, request
from werkzeug.user_agent import UserAgent


class NewUserAgent(UserAgent):
    def new_method(self):
        return f"Hello, world, here is your user agent: {self.string}"


app = Flask(__name__)
app.request_class.user_agent_class = NewUserAgent


@app.route("/")
def index():
    return request.user_agent.new_method()
```

**Reply by SuperSajuuk** [reply]
Thanks! That'll help me a lot 👍
